Framework/Algorithms/inc/MantidAlgorithms/SortXAxis2.h

```diff
@@ -0,0 +1,87 @@
+#ifndef MANTID_ALGORITHMS_SORTXAXIS_H_
+#define MANTID_ALGORITHMS_SORTXAXIS_H_
+
+#include "MantidAPI/Algorithm.h"
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidAPI/WorkspaceProperty.h"
+#include "MantidKernel/ListValidator.h"
+#include "MantidKernel/System.h"
+#include "MantidKernel/make_unique.h"
+
+namespace Mantid {
+namespace Algorithms {
+
+/**
+  Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+
+/**
+ * @brief SortXAxis will take Histogram or Point data and reorder it based on
+ * the X Axis' values, whilst maintaining it's Dx, Y and E axis relative values.
+ * @author Samuel Jones
+ * @version 2.0
+ * @date 07-2018
+ * @copyright GNU General Public License
+ */
+
+class DLLExport SortXAxis : public API::Algorithm {
+public:
+  const std::string name() const override;
+  int version() const override;
+  const std::string category() const override;
+  const std::string summary() const override;
+
+private:
+  void init() override;
+  void exec() override;
+  std::vector<std::size_t> createIndexes(const size_t);
+
+  void sortIndicesByX(std::vector<std::size_t> &workspaceIndecies,
+                      std::string order,
+                      Mantid::API::MatrixWorkspace_const_sptr inputWorkspace,
```

Comment:
It is also generally good practice to avoid passing `shared_ptr`s around between functions where possible. Each copy has to increment the reference count and it also binds the caller to a given memory management style. https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/ is a good read on this. 

I think the workspace parameter can be passed by `const API::MatrixWorkspace &` instead? Similarly for the other cases below?